lib/srv/desktop/rdp/rdpclient/src/lib.rs

```diff
@@ -0,0 +1,359 @@
+#[macro_use]
+extern crate lazy_static;
+
+use libc::{fd_set, select, FD_SET};
+use rdp::core::client::{Connector, RdpClient};
+use rdp::core::event::*;
+use rdp::model::error::*;
+use std::collections::HashMap;
+use std::convert::TryFrom;
+use std::mem;
+use std::net::TcpStream;
+use std::os::unix::io::AsRawFd;
+use std::ptr;
+use std::sync::{Arc, Mutex};
+
+struct Client {
+    rdp_client: RdpClient<TcpStream>,
+    tcp_fd: usize,
+}
+type SyncRdpClient = Arc<Mutex<Client>>;
+
+// Rust-side registry of clients, to allow Go to reference a specific client in calls after
+// connect_rdp.
+lazy_static! {
```

Comment:
Generally globals are heavily discouraged. Does it make more sense to have a Rust object with a pointer passed around C GUI lib style? It can of course make sense to have this but it should be carefully considered.

Reply:
Yeah, this feels gross.
I couldn't figure out how to toss a Rust object pointer over to the Go side, while keeping the object from being cleaned up by Rust's memory management. Also, I'm not sure how to safe it is to reconstruct a Rust object from a raw pointer (all the examples at https://doc.rust-lang.org/1.30.0/book/first-edition/raw-pointers.html deal with primitive types).

Do you know of any docs or examples I could follow?

Reply:
Generally I'd recommend reading the official [nomicon](https://doc.rust-lang.org/nomicon/) as it teaches various dark arts. Rust's memory management is passive and you can use `std::mem::forget` to prevent a struct from being cleaned up. That said the way you want to do this is to `Box<T>` your data object and then call `Box::leak` which as the name suggests, forgets about the memory and gives you a pointer to the value on the heap. You'd then use Go's raw package to deal with this raw pointer. Just don't make it look like a Go value pointer, it needs to be managed as as FFI pointer in Go to prevent it from being GCd. On the Rust side, call `Box::from_raw` to turn the pointer back into a boxed value you can actually use.

Reply:
The more technical term is that structs that manage heap resources such as `Vec` and `Box` implement the Drop trait which acts as a finalizer that runs at end of scope. `mem::forget` prevents that from being run.

Reply:
Also, read the entire nomicon before you continue, it takes a couple hours, it's a great read and prevents you from making many mistakes.

Reply:
That does indeed look correct. But you probably want to give Valgrind, TSAN and MSAN run over this if possible.

Reply:
The invariant that has to be upheld here is that you cannot call multiple functions on the same object concurrent without horrible UB so you might want a lock on the Go side.

Reply:
I added an `Arc<Mutex<...>>` on top of the rust RDP client object: https://github.com/gravitational/teleport/blob/82d382de8f9fbffd318094ebaeb0579c1c574541/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L35
The other field (`tcp_fd`) is read-only.

For releasing memory, I split `free_rdp` away from `close_rdp`.
That way the Go client only releases the Rust memory after all i/o goroutines have finished (after `wg.Wait()` returned).